kernel/src/client.rs

```diff
@@ -0,0 +1,128 @@
+        self.pending_messages.push_back(serialize(msg));
+    }
+
+    fn reconstruct(offset: usize) -> Box<Value> {
+        let ptr = PHYSICAL_ALLOCATOR.from_offset::<Value>(offset);
+        unsafe { Box::from_raw(ptr as *mut Value) }
+    }
+
+    pub fn process_incoming_message(&mut self, msg: Messages, cpu: &mut Cpu) -> bool {
+        match msg {
+            Messages::CreateBlobMessage(m) => {
+                let blob: Blob = unsafe {
+                    Arc::from_raw(core::ptr::slice_from_raw_parts(
+                        PHYSICAL_ALLOCATOR.from_offset::<u8>(m.ptr),
+                        m.size,
+                    ))
+                };
+                self.reply(Box::new(Value::Blob(blob)));
+                log::info!("processed create blob");
+                true
+            }
+            Messages::CreateTreeMessage(m) => {
+                let vals: Box<[usize]> = unsafe {
+                    Box::from_raw(core::ptr::slice_from_raw_parts_mut(
+                        PHYSICAL_ALLOCATOR.from_offset::<usize>(m.ptr) as *mut usize,
+                        m.size,
+                    ))
+                };
+                let mut vec = Vec::new();
+                for v in vals {
+                    vec.push(Box::into_inner(Self::reconstruct(v)));
+                }
+                self.reply(Box::new(Value::Tree(vec.into())));
+                log::info!("processed create tree");
+                true
+            }
+            Messages::CreateThunkMessage(m) => {
+                let v = Self::reconstruct(m.handle);
+                match Box::into_inner(v) {
+                    Value::Blob(b) => self.reply(Box::new(Value::Thunk(Thunk::from_elf(&*b)))),
+                    _ => todo!(),
+                };
+                log::info!("processed create thunk");
+                true
+            }
+            Messages::RunThunkMessage(m) => {
+                let v = Self::reconstruct(m.handle);
+                match Box::into_inner(v) {
+                    Value::Thunk(thunk) => self.reply(Box::new(thunk.run(cpu))),
+                    _ => todo!(),
+                };
+                log::info!("processed run thunk");
+                true
+            }
+            Messages::ApplyMessage(m) => {
+                let v = Box::into_inner(Self::reconstruct(m.lambda_handle));
+                let arg = Box::into_inner(Self::reconstruct(m.arg_handle));
+                match v {
+                    Value::Lambda(lambda) => self.reply(Box::new(Value::Thunk(lambda.apply(arg)))),
+                    _ => todo!(),
+                };
+                log::info!("processed apply lambda");
+                true
+            }
+            Messages::DropMessage(m) => {
+                Self::reconstruct(m.handle);
+                log::info!("processed drop");
+                false
+            }
+            _ => todo!(),
+        }
+    }
```

Comment:
I don't think this logic should be here; either the `Client` is fully responsible for everything (in which case it just has a `run(&mut self, &mut Cpu)` that runs until the client exits), or it's only responsible for communication (in which case this logic should be elsewhere).